packages/amplify_core/lib/src/types/datastore/models/query_snapshot.dart

```diff
@@ -71,37 +71,60 @@ class QuerySnapshot<T extends Model> {
   QuerySnapshot<T> withSubscriptionEvent({
     required SubscriptionEvent<T> event,
   }) {
-    final sortedListCopy = SortedList<T>.from(_sortedList);
     SortedList<T>? updatedSortedList;
 
     final newItem = event.item;
-    final newItemMatchesPredicate = where == null || where!.evaluate(newItem);
-    final currentItemIndex =
+    final matchesPredicate = where == null || where!.evaluate(newItem);
+    final currentIndex =
         // TODO(HuiSF): remove the ignore when merging CPK feature commits
         // ignore: deprecated_member_use_from_same_package
-        sortedListCopy.indexWhere((item) => item.getId() == newItem.getId());
-    final currentItem =
-        currentItemIndex == -1 ? null : sortedListCopy[currentItemIndex];
-    final currentItemMatchesPredicate =
-        currentItem != null && (where == null || where!.evaluate(currentItem));
+        _sortedList.indexWhere((item) => item.getId() == newItem.getId());
+    final currentItem = currentIndex == -1 ? null : _sortedList[currentIndex];
 
-    if (event.eventType == EventType.create &&
-        newItemMatchesPredicate &&
-        currentItem == null) {
-      updatedSortedList = sortedListCopy..addSorted(newItem);
-    } else if (event.eventType == EventType.delete && currentItem != null) {
-      updatedSortedList = sortedListCopy..removeAt(currentItemIndex);
-    } else if (event.eventType == EventType.update) {
-      if (currentItemMatchesPredicate &&
-          newItemMatchesPredicate &&
-          currentItem != newItem) {
-        updatedSortedList = sortedListCopy
-          ..updateAtSorted(currentItemIndex, newItem);
-      } else if (currentItemMatchesPredicate && !newItemMatchesPredicate) {
-        updatedSortedList = sortedListCopy..removeAt(currentItemIndex);
-      } else if (currentItem == null && newItemMatchesPredicate) {
-        updatedSortedList = sortedListCopy..addSorted(newItem);
-      }
+    switch (event.eventType) {
+      case EventType.create:
+        // Skip any new item that doesn't match the predicate.
+        if (!matchesPredicate) break;
+        if (currentItem == null) {
+          // Add the item to the list. This is a new item and matches the
+          // predicate, it should be added.
+          updatedSortedList = _sortedList.copy()..addSorted(newItem);
+        } else if (currentItem != newItem) {
+          // Update the item in the list. This is a "new" item, but it already
+          // exists in the list with a different value. This is the result of
+          // the item being created on this device and App Sync returning an
+          // updated item during the create mutation. This can happen when using
+          // custom resolvers.
+          updatedSortedList = _sortedList.copy()
+            ..updateAtSorted(
+              currentIndex,
+              newItem,
+            );
+        }
```

Comment:
*note:* I refactored this and added comments to make it easier to understand but this code block is the only logic change. When the event type is `EventType.create` and the item value has changed (currentItem != newItem) the item in the list is updated with the new value. Previously this was ignored. This is problem when the model create response from App Sync contains an updated model, which is the case when using a custom resolver.

Reply:
question: is `currentItem != newItem` a sufficient comparison, like is it a deep equals?

Reply:
Yes, item is of type Model and we override the equality operator on models. For example, a typical Blog model would look like:

```dart
  @override
  bool operator ==(Object other) {
    if (identical(other, this)) return true;
    return other is Blog &&
        id == other.id &&
        _name == other._name &&
        DeepCollectionEquality().equals(_posts, other._posts);
  }
```